sdk/nodejs/x/automation/localWorkspace.ts

```diff
@@ -0,0 +1,584 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as fs from "fs";
+import * as yaml from "js-yaml";
+import * as os from "os";
+import * as upath from "upath";
+
+import { CommandResult, runPulumiCmd } from "./cmd";
+import { ConfigMap, ConfigValue } from "./config";
+import { ProjectSettings } from "./projectSettings";
+import { Stack } from "./stack";
+import { StackSettings } from "./stackSettings";
+import { PulumiFn, StackSummary, WhoAmIResult, Workspace } from "./workspace";
+
+/**
+ * LocalWorkspace is a default implementation of the Workspace interface.
+ * A Workspace is the execution context containing a single Pulumi project, a program,
+ * and multiple stacks. Workspaces are used to manage the execution environment,
+ * providing various utilities such as plugin installation, environment configuration
+ * ($PULUMI_HOME), and creation, deletion, and listing of Stacks.
+ * LocalWorkspace relies on Pulumi.yaml and Pulumi.<stack>.yaml as the intermediate format
+ * for Project and Stack settings. Modifying ProjectSettings will
+ * alter the Workspace Pulumi.yaml file, and setting config on a Stack will modify the Pulumi.<stack>.yaml file.
+ * This is identical to the behavior of Pulumi CLI driven workspaces.
+ *
+ * @alpha
+ */
+export class LocalWorkspace implements Workspace {
+    /**
+     * The working directory to run Pulumi CLI commands
+     */
+    readonly workDir: string;
+    /**
+     * The directory override for CLI metadata if set.
+     * This customizes the location of $PULUMI_HOME where metadata is stored and plugins are installed.
+     */
+    readonly pulumiHome?: string;
+    /**
+     * The secrets provider to use for encryption and decryption of stack secrets.
+     * See: https://www.pulumi.com/docs/intro/concepts/config/#available-encryption-providers
+     */
+    readonly secretsProvider?: string;
+    /**
+     *  The inline program `PulumiFn` to be used for Preview/Update operations if any.
+     *  If none is specified, the stack will refer to ProjectSettings for this information.
+     */
+    public program?: PulumiFn;
```

Comment:
Nit: `public` is the default visibility. It would be nice to be consistent whether it's used or not.

Reply:
Removed public declarations in favor of implicit defaults.